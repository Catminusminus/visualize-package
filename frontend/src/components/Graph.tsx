/**
 * Author of this code work, Yuki Mogi. c FFRI, Inc. 2019
 */

import * as React from 'react'
import Chart from 'react-google-charts'
import Paper from '@material-ui/core/Paper'
import { withStyles, WithStyles, createStyles } from '@material-ui/core/styles'

const styles = theme =>
  createStyles({
    root: {
      display: 'flex',
      justifyContent: 'center',
      flexWrap: 'wrap',
      padding: theme.spacing(1 / 2)
    }
  })

interface Props extends WithStyles<typeof styles> {
  title: string
  data: {} | any[]
}

const Graph: React.FC<Props> = ({ title, data, classes }) => {
  return (
    <Paper className={classes.root}>
      <Chart
        width={'500px'}
        height={'400px'}
        chartType="LineChart"
        loader={<div>Drawing Chart...</div>}
        data={data}
        options={{
          title,
          vAxis: {
            title: 'Number'
          },
          hAxis: {
            title: 'Time'
          },
          explorer: {
            axis: 'horizontal',
            actions: ['dragToZoom', 'rightClickToReset']
          }
        }}
        rootProps={{ 'data-testid': '2' }}
      />
    </Paper>
  )
}

export default withStyles(styles)(Graph)
