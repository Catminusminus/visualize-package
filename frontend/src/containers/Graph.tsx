/**
 * Author of this code work, Yuki Mogi. c FFRI, Inc. 2019
 */

import * as React from 'react'
import GraphComponent from '../components/Graph'
import { useSelector } from 'react-redux'
import { State } from '../modules'
import Grid from '@material-ui/core/Grid'
import {
  makeGraphDataFromStrings,
  makeGraphDataFromObjects
} from '../utils/MakeGraphData'

const packageInfoSelector = (state: State) => ({
  packageUpdates: state.reducer.packageUpdates,
  commits: state.reducer.commits,
  openIssues: state.reducer.openIssues,
  openPRs: state.reducer.openPRs
})

const Graph = () => {
  const { packageUpdates, commits, openIssues, openPRs } = useSelector(
    packageInfoSelector
  )
  const updates = makeGraphDataFromStrings(packageUpdates)
  const commitList = makeGraphDataFromStrings(commits)
  const issues = makeGraphDataFromObjects(openIssues)
  const pulls = makeGraphDataFromObjects(openPRs)
  type SafeGraphProps = {
    title: string
    data: string[][]
  }
  /*
    GraphComponents with 2*2 data will generate a error.
    ex)
    <GraphComponent data={
      [
        ['x', 'y'],
        [new Date(), 0]
      ]
    } />
  */
  const SafeGraph = ({ title, data }: SafeGraphProps) => (
    <>
      {data.length > 2 ? (
        <GraphComponent
          title={title}
          data={
            data.toString() === [['x']].toString()
              ? [
                  ['x', 'y'],
                  [0, 0]
                ]
              : data
          }
        />
      ) : (
        <div>Too Few Data To Draw A Chart</div>
      )}
    </>
  )

  return (
    <>
      <Grid item xs={12} md={6}>
        <Grid container justify="center">
          <SafeGraph title="Open PRs" data={pulls} />
        </Grid>
      </Grid>
      <Grid item xs={12} md={6}>
        <Grid container justify="center">
          <SafeGraph title="Open Issues" data={issues} />
        </Grid>
      </Grid>
      <Grid item xs={12} md={6}>
        <Grid container justify="center">
          <SafeGraph title="Commits" data={commitList} />
        </Grid>
      </Grid>
      <Grid item xs={12} md={6}>
        <Grid container justify="center">
          <SafeGraph title="Package Updates" data={updates} />
        </Grid>
      </Grid>
    </>
  )
}

export default Graph
