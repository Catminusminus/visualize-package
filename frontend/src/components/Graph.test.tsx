/**
 * Author of this code work, Yuki Mogi. c FFRI, Inc. 2019
 */

import * as React from 'react'
import renderer from 'react-test-renderer'
import Graph from './Graph'

it('Snapshot test of Graph', () => {
  const result = renderer
    .create(
      <Graph
        data={[
          ['x', 'y'],
          [new Date('2018/09'), 10],
          [new Date('2018/10'), 5]
        ]}
        title="Chart"
      />
    )
    .toJSON()
  expect(result).toMatchSnapshot()
})
