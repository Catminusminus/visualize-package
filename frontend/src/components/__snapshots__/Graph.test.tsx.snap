// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test of Graph 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 Graph-root-1 MuiPaper-rounded"
>
  <div>
    Drawing Chart...
  </div>
</div>
`;
